Jetty-12 Improved Content testing and javadoc

gregw opened the pull request:
Some javadoc improvements and some minor refinements

gregw committed:
Improved Content javadoc

Some javadoc improvements and some minor refinements

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-content-javadoc

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -462,6 +463,20 @@ static Chunk from(ByteBuffer byteBuffer, boolean last, Runnable releaser)
+        static Chunk from(ByteBuffer byteBuffer, boolean last, Consumer<ByteBuffer> releaser)
+        {
+            // TODO make this non capturing
+            return new ByteBufferChunk(byteBuffer, last, () -> releaser.accept(byteBuffer));
```
I think you should add a null check for `releaser`.

gregw replied:
I've instead implemented the non capturing approach, which handles null

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceInputStream.java:
```diff
@@ -21,6 +21,13 @@
+ * <p>
+ * The read methods are implemented by calling {@link Content.Source#read()}.
+ * Any {@link Content.Chunk}s read are released once all their content
+ * has been read.
```
I have a feeling you just put your finger on something important we missed: what should happen when the `InputStream` is closed?

IIRC some (all?) `Content.Source` implementations expect that `read()` be called continuously until a last chunk is returned, or that `fail()` is called, otherwise some resources might be leaked. `PathContentSource` for instance is impacted by this problem.

gregw replied:
We are correctly releasing the chunk on a close, but I think we should set the next chunk to an `Error` if we closed with a non null non error chunk.

lorban replied:
You are correctly releasing the `Chunk` contained in `ContentSourceInputStream` but you're not notifying the wrapped `Content.Source` to release its chunks.

gregw replied:
@lorban I'm not sure there is any compulsion to read all the content from a Source, so close not reading to completion is not a big deal.
However, I have pushed a version  that if close is done and EOF has not been read, then not only is the chunk set to an Error, but the source is failed.
@sbordet your thoughts?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceTransformer.java:
```diff
@@ -17,6 +17,17 @@
+ * <p>
+ * The {@link #demand(Runnable)} conversation is passed directly to the wrapped {@link Content.Source}, which means
+ * that transformations that may fully consume bytes read can result in a null return from {@link Content.Source#read()}
+ * even after a callback to the demand {@link Runnable}.
```
This last paragraph is (should?) already be covered by the fact that `demand()` may lead to spurious callbacks. I'd either remove this paragraph or link to the one that speaks about `demand()`'s spurious behavior.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceTransformer.java:
```diff
@@ -112,5 +123,22 @@ private Content.Chunk process(Content.Chunk rawChunk)
+     * <li>the <code>rawChunk</code>. This is typically done for {@link Content.Chunk.Error}s,
+     *     when {@link Content.Chunk#isLast()} is true, or if no transformation is required.</li>
+     * <li>a new (or predefined) {@link Content.Chunk} derived from the <code>rawChunk</code>. The transform is
+     *     responsible for calling {@link Content.Chunk#release()} on the <code>rawChink</code>, either during the call
```
`<code>rawChink</code>` ?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/InputStreamContentSource.java:
```diff
@@ -16,24 +16,42 @@
+ * A {@link Content.Source} that is backed by an {@link InputStream}.
+ * <p>
+ * Data is read from the {@link InputStream} into a buffer that is optionally acquired
+ * from a {@link ByteBufferPool}, and converted to a {@link Content.Chunk} that is
```
You should specify where the buffer pool is coming from when none is given during construction.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/OutputStreamContentSource.java:
```diff
@@ -21,6 +21,14 @@
 import org.eclipse.jetty.util.IO;
 
+/**
+ * A {@link Content.Source} backed backed by an {@link OutputStream}.
```
`backed backed` ?

lorban reviewed (changes requested):
Reading the english-written contracts makes it more apparent that there are things we overlooked and others that need to be improved.

The [MultiPart PR](https://github.com/eclipse/jetty.project/pull/8201) contains some related changes that probably should be reviewed together with this PR.

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-content-javadoc

gregw committed:
updates from review

gregw commented:
@lorban @sbordet Can I get this (re)reviewed please?  I've reviewed the multipart changes to the javadoc and they are orthogonal/complimentary.

lorban reviewed (changes requested):
The `ContentSourceInputStream` does not notify the wrapped `Content.Source` when it is closed, potentially making it leak its internal resources. @sbordet it feels like something is unclear about the `Content.Source` contract that makes this problem non-obvious.

gregw committed:
close fails content

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceInputStream.java:
```diff
@@ -89,10 +96,21 @@ public int available() throws IOException
+            if (!chunk.isTerminal())
+                chunk.release();
+            return;
+        }
```
What? We certainly don't want to do this!

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceInputStream.java:
```diff
@@ -89,10 +96,21 @@ public int available() throws IOException
-        if (chunk == Content.Chunk.EOF || chunk instanceof Content.Chunk.Error)
-            return;
-        chunk = Content.Chunk.EOF;
+
+        Throwable closed = new IOException("closed before EOF");
+        chunk = Content.Chunk.from(closed);
+        content.fail(closed);
```
```suggestion
        if (chunk == Content.Chunk.EOF || chunk instanceof Content.Chunk.Error)
            return;
        Throwable closed = new IOException("closed before EOF");
        chunk = Content.Chunk.from(closed);
        content.fail(closed);
```
I think a combination of old code and new code is better.

gregw replied:
@sbordet the problem with your version is that it tests for `== EOF`, thus it will not see a normal chunk with isLast()==true as end of file and will terminate with a failure.

So my version is trying to say, that if a last chunk has been read, then a close is perfectly OK even if that chunk has unconsumed content... we just have to release it.   
I'll try to make that more explicit....

gregw replied:
Ah I didn't see that isTerminal checks for no remaining bytes, so my code is kind of wrong?   Or maybe not, should closing a non terminal last chunk be a failure?

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/InputStreamContentSource.java:
```diff
@@ -59,17 +78,17 @@ public Content.Chunk read()
-            byte[] buffer = new byte[getBufferSize()];
-            int read = inputStream.read(buffer);
+            ByteBuffer buffer = bufferPool.acquire(getBufferSize(), false);
+            int read = inputStream.read(buffer.array());
```
Not sure you can do this... Shouldn't `arrayOffset`, etc. be used?

gregw replied:
technically yes, but currently no buffer allocated by our pools will have a non zero arrayOffset.  But I will add.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/InputStreamContentSource.java:
```diff
@@ -59,17 +78,17 @@ public Content.Chunk read()
             {
-                ByteBuffer byteBuffer = ByteBuffer.wrap(buffer, 0, read);
-                return Content.Chunk.from(byteBuffer, false);
+                buffer.position(read);
```
Buffer should be flipped and sliced.

gregw replied:
why is slice needed?  The buffer is not retained nor reused by this class, so once the consumer calls release it is put back into the pool.

Perhaps it could be held expecting a release?  But this is simpler for now unless there is a performance indication that we need to do better.

gregw replied:
Hmmm ContentSourceTest tests all the failure modes, but does test actually reading the content!  So this code is a) wrong; b) not tested!  fixing....

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/AsyncContentSourceTest.java:
```diff
@@ -53,7 +59,7 @@ public void testCloseInvokesDemandCallback() throws Exception
         CountDownLatch latch = new CountDownLatch(1);
         async.demand(latch::countDown);
-        assertFalse(latch.await(500, TimeUnit.MILLISECONDS));
+        assertFalse(latch.await(100, TimeUnit.MILLISECONDS));
```
Don't think only 100 ms is a good idea.

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/AsyncContentSourceTest.java:
```diff
@@ -67,27 +73,52 @@ public void testCloseInvokesDemandCallback() throws Exception
-        assertFalse(latch.await(500, TimeUnit.MILLISECONDS));
+            CountDownLatch latch = new CountDownLatch(1);
+            async.demand(latch::countDown);
+            assertFalse(latch.await(100, TimeUnit.MILLISECONDS));
```
Too short.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java:
```diff
@@ -683,5 +683,21 @@ public Request.Processor handle(Request request) throws Exception
+            {
+                super(InvocationType.NON_BLOCKING);
+            }
+        }
```
Why is this part of this PR?

gregw replied:
Ah it wasn't meant to be... but kind of is OK.  I'm working on a blog/tutorial and thus fixing missing documentation and making better examples.  Having these inner classes greatly improves those examples and usability.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/Invocable.java:
```diff
@@ -13,6 +13,8 @@
 package org.eclipse.jetty.util.thread;
 
+import org.eclipse.jetty.util.Callback;
+
```
Unused import?

gregw replied:
Is used in javadoc, but I have made that fully qualified instead of an import.

sbordet reviewed (changes requested):
Have you confronted this PR with the multipart PR, that has more extensive javadocs?
Why the need for this PR when the multipart was already doing it?

gregw committed:
Updates from review

Fixed ContentSourceTest that was consuming the same ByteBufferSource multiple times.

gregw committed:
Updates from review

Added back tests for OutputStreamContentSource

gregw commented:
The ContentSourceTests were not testing very much and had errors.  I have improved the tests and fixed some bugs, but there are still some that need to be checked.

gregw committed:
Updates from review

Fixed name clash on Blocker vs Blocking

gregw committed:
Updates from review

tutorial example

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-content-javadoc

# Conflicts:
#	jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/EmptyServerHandler.java
#	jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSinkOutputStream.java
#	jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/EchoHandler.java

gregw committed:
Improve javadoc style

gregw committed:
Simplified test

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPool.java:
```diff
@@ -27,6 +27,8 @@
  */
 public interface ByteBufferPool
 {
+    ByteBufferPool NULL_POOL = new NullByteBufferPool();
```
I'd prefer this to be called `NOOP_POOL` and `NoopByteBufferPool`, as "null" does not really convey what's doing.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -223,6 +224,38 @@ static void consumeAll(Source source, Callback callback)
+            {
+                throw new RuntimeException(e);
+            }
+        }
```
I don't like this infinitely-waiting method that is only used by tests be so public in the main interface.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -459,7 +492,24 @@ static Chunk from(ByteBuffer byteBuffer, boolean last)
-            return new ByteBufferChunk(byteBuffer, last, releaser);
+            return releaser == null
+                ? new ByteBufferChunk(byteBuffer, last)
+                : new ByteBufferChunk.ReleasedByRunnable(byteBuffer, last, releaser);
```
Now that the non-releaser version directly allocates, I would mandate that the 3rd parameter must be non-null.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -459,7 +492,24 @@ static Chunk from(ByteBuffer byteBuffer, boolean last)
+        {
+            return releaser == null
+                ? new ByteBufferChunk(byteBuffer, last)
+                : new ByteBufferChunk.ReleasedByConsumer(byteBuffer, last, releaser);
```
Now that the non-releaser version directly allocates, I would mandate that the 3rd parameter must be non-null.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/NullByteBufferPool.java:
```diff
@@ -21,6 +21,9 @@ public class NullByteBufferPool implements ByteBufferPool
 
+    public NullByteBufferPool()
+    {}
+
```
We don't need this (it's generated automatically), unless it was meant to be `private` since there is now the constant in the interface.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java:
```diff
@@ -27,6 +27,13 @@
+/**
+ * <p>A {@link Content.Source} that is also a {@link Content.Sink}.
+ * Content written to the {@link Content.Sink} is converted to a {@link Content.Chunk}
+ * and made available to calls to the {@link #read()} method.  In necessary, any
```
`In` -> `If`.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java:
```diff
@@ -27,6 +27,13 @@
+ * Content written to the {@link Content.Sink} is converted to a {@link Content.Chunk}
+ * and made available to calls to the {@link #read()} method.  In necessary, any
+ * {@link Runnable} passed to the {@link #demand(Runnable)} method is invoked once
+ * content is written to the {@link Content.Sink}.</p?
```
`</p?` typo.

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-content-javadoc

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceInputStream.java:
```diff
@@ -89,10 +96,26 @@ public int available() throws IOException
+        // This is an abnormal close before EOF
+        Throwable closed = new IOException("closed before EOF");
+        chunk = Content.Chunk.from(closed);
+        content.fail(closed);
```
Still think we should not do any `read()` from `close()`.
Previous logic plus the last three lines above to fail before EOF.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceTransformer.java:
```diff
@@ -112,5 +124,22 @@ private Content.Chunk process(Content.Chunk rawChunk)
 
+    /**
+     * Content chunk transformation method.
+     * <p>
```
Improve `<p>` wrapping (one per paragraph).

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/InputStreamContentSource.java:
```diff
@@ -59,17 +77,17 @@ public Content.Chunk read()
-            byte[] buffer = new byte[getBufferSize()];
-            int read = inputStream.read(buffer);
+            ByteBuffer buffer = bufferPool.acquire(getBufferSize(), false);
+            int read = inputStream.read(buffer.array(), buffer.arrayOffset(), buffer.capacity());
```
```suggestion
            int read = inputStream.read(buffer.array(), buffer.arrayOffset(), buffer.remaining());
```
This is more correct. We have `MappedByteBufferPool.Tagged` that slices the buffers of the pool, so `remaining()` is correct, not `capacity()`.

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/AsyncContentSourceTest.java:
```diff
@@ -32,18 +37,19 @@
     public void testOfferInvokesDemandCallback() throws Exception
     {
-        AsyncContent async = new AsyncContent();
+        try (AsyncContent async = new AsyncContent())
```
No need of try-with-resources.

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/AsyncContentSourceTest.java:
```diff
@@ -32,18 +37,19 @@
+        {
+            CountDownLatch latch = new CountDownLatch(1);
+            async.demand(latch::countDown);
+            assertFalse(latch.await(250, TimeUnit.MILLISECONDS));
```
Beats me why you want to use 250 ms given all the slowness of CI. I'd restore at least 500.

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ContentSourceTest.java:
```diff
@@ -65,15 +69,21 @@ public static List<Content.Source> all() throws Exception
+        ByteBufferContentSource byteBufferSource = new ByteBufferContentSource(UTF_8.encode("one"), UTF_8.encode("two"));
 
-        Content.Source.Transformer transformerSource = new Content.Source.Transformer(byteBufferSource)
+        Content.Source.Transformer transformerSource = new Content.Source.Transformer(new ByteBufferContentSource(UTF_8.encode("one"), UTF_8.encode("two")))
```
Duplicated, revert to old code.

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ContentSourceTest.java:
```diff
@@ -65,15 +69,21 @@ public static List<Content.Source> all() throws Exception
+            @Override
+            public String toString()
+            {
+                return "transformerSource@%x".formatted(hashCode());
```
Please no, let's use `Class.simpleName` instead of hardcoded strings that don't refactor in case of renamings.

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ContentSourceTest.java:
```diff
@@ -83,31 +93,99 @@ protected Content.Chunk transform(Content.Chunk rawChunk)
+            @Override
+            public String toString()
+            {
+                return "OutputStreamContentSource@%x".formatted(hashCode());
```
Use `Class.simpleName`.

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ContentSourceTest.java:
```diff
@@ -83,31 +93,99 @@ protected Content.Chunk transform(Content.Chunk rawChunk)
     @ParameterizedTest
     @MethodSource("all")
-    public void testDemandReadDemandDoesNotRecurse(Content.Source source)
+    public void testRead(Content.Source source) throws Exception
```
Rename to `testReadToEOF()`.

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ContentSourceTest.java:
```diff
@@ -121,21 +199,19 @@ public void run()
             }
         });
-        assertTrue(processed.get());
+        assertTrue(processed.await(10, TimeUnit.SECONDS));
```
The previous use of `AtomicBoolean` made this test stricter on purpose. Using a `CountDownLatch` is not exactly the same (there was no need to change it).

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ContentSourceTest.java:
```diff
@@ -146,16 +222,17 @@ public void run()
             }
         });
-        assertTrue(processed.get());
+        assertTrue(processed.await(10, TimeUnit.SECONDS));
```
The previous use of `AtomicBoolean` made this test stricter on purpose. Using a `CountDownLatch` is not exactly the same (there was no need to change it).

gregw replied:
Without the latch, the test is assuming that the demand callback will be called within the scope of the demand call. That's not necessarily true in general, even if content is available.  It is definitely not true for OutputStreamContentSource, which uses a different thread to add the content after a demand call.

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ContentSourceTest.java:
```diff
@@ -146,16 +222,17 @@ public void run()
+    public void testReadFailReadReturnsError(Content.Source source) throws Exception
     {
-        Content.Chunk chunk = source.read();
+        Content.Chunk chunk = Content.Source.nextChunk(source);
```
Why this? All sources should return a non null initial chunk all the times.

gregw replied:
@sbordet No the OutputStreamContentSource doesn't

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ContentSourceTest.java:
```diff
@@ -199,9 +271,13 @@ public void testReadErrorDemandInvokesDemandCallback(Content.Source source) thro
+        if (source instanceof OutputStreamContentSource)
+            return;
+
+        Content.Chunk chunk = Content.Source.nextChunk(source);
```
Why this? All sources should return a non null initial chunk all the times.

gregw replied:
OutputStreamContentSource doesn't

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -1005,12 +1005,16 @@ public boolean content(ByteBuffer buffer)
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("release {}/{} for {}", BufferUtil.toDetailString(buffer), retainable, this);
+                })
+                : Content.Chunk.from(buffer, false, retainable::release);
```
I'd revert to old code, as the lambda is allocating anyway. `retainable::release` is nice to write, but it's no different from `() -> retainable.release()` which captures `retainable` which allocates.

gregw replied:
good point

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/TutorialTest.java:
```diff
@@ -0,0 +1,196 @@
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.thread.SerializedInvoker;
+
+public class TutorialTest
```
No. This should be converted to a documentation class, see all examples in the documentation, e.g. `ServerDocs.java`.

gregw replied:
Thanks, as the comment said... it was looking for a good home... but documentation is not currently being built and needs a bit of work to update it to 12 layout.

gregw committed:
reverted change

gregw committed:
new home

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Updates from review

sbordet committed:
Using ByteBufferPool.NOOP instead of allocating NoopByteBufferPool.
Removed TutorialTest.java (moved to HandlerDocs.java).
Removed unnecessary override in OutputStreamContentSource.java.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
